File: inspecto-ui/src/app/inspecto/graph/branching-pattern-engine.ts
import type { G6Edge, G6GraphData } from './graph-types';
import type { GraphSelection } from './graph-analysis';
import { analysisNodeCapValue, baseEdgeKind, edgeTimeIndex, followsInTime } from './graph-analysis';

/**
 * **Branching pattern runtime (LA-14b)** — fan-out / fan-in motifs over the shared {@link G6GraphData} shape.
 *
 * `matchPattern` finds LINEAR motifs: one node per step, one edge per hop. That cannot express the shapes
 * money actually takes when it is being hidden — **structuring** (many sub-threshold deposits converging on
 * one collector, which then splits the total across several intermediaries that re-converge on one exit)
 * is a set of parallel legs at each stage, not a path. This file EXTENDS the linear matcher rather than
 * replacing it: linear packs still run through `matchPattern`, and this matcher reuses its temporal rule
 * ({@link followsInTime}), its time parse ({@link edgeTimeIndex}) and its kind folding ({@link baseEdgeKind}),
 * so the two can never disagree about when a link happened or what kind it is.
 *
 * <p><b>The model.</b> A motif is an ordered list of {@link BranchStage}s. The match carries a FRONTIER —
 * the nodes it has reached, each with the time it reached them.
 * <ul>
 *   <li><b>fan-in</b> at stage 0: every node with ≥ `minBranches` distinct counterparties sending it a
 *       qualifying leg is a candidate collector; the frontier becomes `{collector}`.</li>
 *   <li><b>fan-out</b> at stage 0: every node sending qualifying legs to ≥ `minBranches` distinct nodes is a
 *       candidate splitter; the frontier becomes those targets.</li>
 *   <li><b>fan-out</b> later: the frontier splits into ≥ `minBranches` distinct new nodes; they become the
 *       frontier.</li>
 *   <li><b>fan-in</b> later (re-convergence): ≥ `minBranches` distinct frontier members each send a leg to
 *       one common new node. Every such node is a separate match.</li>
 * </ul>
 * A leg into or out of a frontier node obeys LA-14a's rule against the time the match reached THAT node —
 * per branch, not per stage — so relay-01 forwarding before it was paid is rejected even when relay-02's
 * timing is fine. The time at which a fan-in "arrives" is the moment it reached its minimum breadth (the
 * `minBranches`-th distinct counterparty), which is what "collected, then moved on" means.
 *
 * <p>⛔ <b>Fails closed and says why</b>, the way LA-14a does — {@link BranchingResult.refusal} is set,
 * never an empty result that reads as "not present":
 * a temporal motif with no time column · a threshold on an attribute no link carries · a threshold that
 * NO link in the graph passes (the plan §2.6 filter trap: a view filtered to `AMOUNT ≥ 5 000` has already
 * removed every structuring leg before this runs, and "no matches" would declare the structuring absent) ·
 * a graph over the analysis cap. None of these throws: the cap is a refusal, not an exception.
 *
 * <p><b>Bounded.</b> Every leg examination counts against {@link BRANCHING_WORK_BUDGET}; on breach the
 * search stops and {@link BranchingResult.truncated} is set, so it can never hang the main thread.
 */

/**
 * A per-leg numeric range — the pack's VISIBLE threshold, e.g. `900 ≤ AMOUNT < 1000`. `min` is inclusive and
 * `max` exclusive, because a reporting threshold is crossed AT its value: a 1 000 deposit is reported, a
 * 999.99 one is not. Either bound may be absent. A band (both bounds) is what "just under the threshold"
 * means, and it is what keeps ordinary small payments — the long tail under any threshold — out of the motif.
 */
export interface LegThreshold {
    /** The edge attribute column (entity-projection `attrCols`), parsed with `Number`. */
    attr: string;
    min?: number;
    max?: number;
}

/** One stage of a branching motif. The temporal fields are LA-14a's, with the same meaning. */
export interface BranchStage {
    shape: 'fan-in' | 'fan-out';
    /** Minimum DISTINCT counterparties at this stage (senders for fan-in, receivers for fan-out). ≥ 1. */
    minBranches: number;
    /** The kind each leg must be (base kind); wildcard when absent. */
    edgeKind?: string;
    /** The kind of the node(s) this stage reaches (the collector for fan-in, each branch for fan-out). */
    nodeKind?: string;
    /** Each counted leg must pass this — legs that fail it are not part of the motif. */
    threshold?: LegThreshold;
    /** All counted legs of this stage fall within this many hours of each other (per collector/splitter). */
    windowHours?: number;
    /** Each leg is strictly after the match reached its tail node. Ignored on stage 0. */
    afterPrevious?: boolean;
    /** Upper bound on that gap, in hours. Only meaningful with {@link afterPrevious}. */
    maxGapHours?: number;
}

export interface BranchingOptions {
    /** The edge attribute holding the event time — the pane's time column, exactly as for `matchPattern`. */
    timeAttr?: string;
    /** Maximum matches returned (default 200). */
    limit?: number;
}

/** One match: every node and leg of the motif, plus the node LAYERS in stage order for a readable label. */
export interface BranchingMatch extends GraphSelection {
    /** e.g. `[[smurf…], [hub], [relay-01, relay-02], [offshore]]` for structuring. */
    layers: string[][];
}

export interface BranchingResult {
    matches: BranchingMatch[];
    /** Set when the motif could not be evaluated — the reason, in the analyst's words. Matches are empty. */
    refusal?: string;
    /** The work budget or the match limit was hit, so there may be more matches than returned. */
    truncated: boolean;
}

/** Ceiling on leg examinations per run. ~6 000 legs × a handful of stages is far below it. */
export const BRANCHING_WORK_BUDGET = 500_000;

/** Whether a motif needs a time column to mean anything — a window, or ordering after stage 0. */
export function branchingNeedsTime(stages: BranchStage[]): boolean {
    return stages.some((st, i) => (st.windowHours ?? 0) > 0 || (i > 0 && !!st.afterPrevious));
}

/** A threshold rendered the way the analyst reads it, e.g. `900 ≤ AMOUNT < 1000`. */
export function thresholdLabel(t: LegThreshold): string {
    return [t.min !== undefined ? `${t.min} ≤` : '', t.attr, t.max !== undefined ? `< ${t.max}` : '']
        .filter(Boolean)
        .join(' ');
}

function passes(t: LegThreshold, v: number): boolean {
    return (t.min === undefined || v >= t.min) && (t.max === undefined || v < t.max);
}

function numericAttr(e: G6Edge, attr: string): number | undefined {
    const raw = e.data.attrs?.[attr];
    if (raw == null || String(raw).trim() === '') return undefined;
    const v = Number(raw);
    return Number.isFinite(v) ? v : undefined;
}

/** A leg in the frame of the stage node it is grouped under: `other` is the counterparty. */
interface Leg {
    edgeId: string;
    other: string;
    t: number | undefined;
}

/**
 * The legs a stage counts for one collector/splitter, reduced to the EARLIEST window holding ≥ `min` distinct
 * counterparties (or all of them when there is no window). Returns null when the breadth is not reached.
 * `arrival` is the time the `min`-th distinct counterparty appeared — when the fan "completed".
 */
function selectWindow(
    legs: Leg[],
    min: number,
    windowHours: number | undefined,
): { legs: Leg[]; arrival: number | undefined } | null {
    const distinct = new Set(legs.map((l) => l.other));
    if (distinct.size < min) return null;
    const timed = legs.filter((l) => l.t !== undefined).sort((a, b) => a.t! - b.t!);
    const completion = (ordered: Leg[]): number | undefined => {
        const seen = new Set<string>();
        for (const l of ordered) {
            seen.add(l.other);
            if (seen.size >= min) return l.t;
        }
        return undefined;
    };
    if (!windowHours || windowHours <= 0) return { legs, arrival: completion(timed) };
    const span = windowHours * 3_600_000;
    for (let i = 0, j = 0; i < timed.length; i++) {
        while (j < timed.length && timed[j].t! - timed[i].t! <= span) j++;
        const win = timed.slice(i, j);
        if (new Set(win.map((l) => l.other)).size >= min) return { legs: win, arrival: completion(win) };
    }
    return null;
}

/** A fan-out's new frontier: each reached node with the EARLIEST time a counted leg reached it. */
function earliestPerNode(legs: Leg[]): Map<string, number | undefined> {
    const next = new Map<string, number | undefined>();
    for (const l of legs) {
        const prev = next.get(l.other);
        if (!next.has(l.other) || (l.t !== undefined && (prev === undefined || l.t < prev))) next.set(l.other, l.t);
    }
    return next;
}

/**
 * Find every match of a branching motif. See the file header for the model and the refusals.
 * Never mutates its input and never throws on a well-typed motif.
 */
export function matchBranchingPattern(
    g: G6GraphData,
    stages: BranchStage[],
    opts: BranchingOptions = {},
): BranchingResult {
    const { timeAttr, limit = 200 } = opts;
    const refuse = (refusal: string): BranchingResult => ({ matches: [], refusal, truncated: false });
    if (!stages.length) return refuse('This pattern has no stages.');
    const cap = analysisNodeCapValue();
    if (g.nodes.length > cap) {
        return refuse(`Branching patterns are capped at ${cap} nodes (graph has ${g.nodes.length}).`);
    }
    if (branchingNeedsTime(stages) && !timeAttr) {
        return refuse('This pattern has a time window or ordering — choose a time column in the Query panel first.');
    }

    const nodeKind = new Map(g.nodes.map((n) => [n.id, n.data.kind]));
    const kindOk = (id: string, k?: string): boolean => !k || nodeKind.get(id) === k;
    const edgeKind = new Map(g.edges.map((e) => [e.id, baseEdgeKind(e.data.kind)]));
    const edgeTime = edgeTimeIndex(g, timeAttr);

    // Per stage: the legs it may count, after kind + threshold. A threshold nothing passes is the §2.6 trap.
    const eligible: G6Edge[][] = [];
    for (const st of stages) {
        let legs = g.edges.filter((e) => e.source !== e.target && (!st.edgeKind || edgeKind.get(e.id) === st.edgeKind));
        if (st.threshold) {
            const th = st.threshold;
            const valued = legs.filter((e) => numericAttr(e, th.attr) !== undefined);
            if (!valued.length) {
                return refuse(
                    `No link carries a numeric ${th.attr}, so the threshold ${thresholdLabel(th)} cannot be evaluated — add ${th.attr} as a link attribute in the Query panel.`,
                );
            }
            legs = valued.filter((e) => passes(th, numericAttr(e, th.attr)!));
            if (!legs.length) {
                return refuse(
                    `No link in this graph passes ${thresholdLabel(th)}. If the view filters ${th.attr} (for example to ≥ 5 000), the legs this pattern looks for were removed before it ran — clear that filter rather than read this as "none found".`,
                );
            }
        }
        eligible.push(legs);
    }
    const index = eligible.map((legs) => {
        const out = new Map<string, Leg[]>();
        const inn = new Map<string, Leg[]>();
        for (const e of legs) {
            const t = edgeTime.get(e.id);
            if (!out.has(e.source)) out.set(e.source, []);
            out.get(e.source)!.push({ edgeId: e.id, other: e.target, t });
            if (!inn.has(e.target)) inn.set(e.target, []);
            inn.get(e.target)!.push({ edgeId: e.id, other: e.source, t });
        }
        return { out, in: inn };
    });

    let work = 0;
    let truncated = false;
    const spend = (n: number): boolean => {
        work += n;
        if (work > BRANCHING_WORK_BUDGET) truncated = true;
        return !truncated;
    };
    const matches: BranchingMatch[] = [];

    interface State {
        frontier: Map<string, number | undefined>;
        used: Set<string>;
        edges: string[];
        layers: string[][];
    }

    const extend = (stageIdx: number, s: State): void => {
        if (truncated || matches.length >= limit) return;
        if (stageIdx >= stages.length) {
            matches.push({ nodeIds: [...s.used], edgeIds: [...new Set(s.edges)], layers: s.layers });
            return;
        }
        const st = stages[stageIdx];
        const { out } = index[stageIdx];
        if (st.shape === 'fan-out') {
            // The frontier splits: legs out of every frontier node, each ordered against ITS arrival.
            const legs: Leg[] = [];
            for (const [f, at] of s.frontier) {
                const cand = out.get(f) ?? [];
                if (!spend(cand.length)) return;
                for (const l of cand) {
                    if (s.used.has(l.other) || !kindOk(l.other, st.nodeKind)) continue;
                    if (followsInTime(at, l.t, st)) legs.push(l);
                }
            }
            const win = selectWindow(legs, st.minBranches, st.windowHours);
            if (!win) return;
            const next = earliestPerNode(win.legs);
            extend(stageIdx + 1, {
                frontier: next,
                used: new Set([...s.used, ...next.keys()]),
                edges: [...s.edges, ...win.legs.map((l) => l.edgeId)],
                layers: [...s.layers, [...next.keys()]],
            });
            return;
        }
        // fan-in (re-convergence): candidate collectors are the new nodes frontier members send to.
        const byCollector = new Map<string, Leg[]>();
        for (const [f, at] of s.frontier) {
            const cand = out.get(f) ?? [];
            if (!spend(cand.length)) return;
            for (const l of cand) {
                if (s.used.has(l.other) || !kindOk(l.other, st.nodeKind) || !followsInTime(at, l.t, st)) continue;
                if (!byCollector.has(l.other)) byCollector.set(l.other, []);
                // Re-framed so `other` is the SENDER — breadth is counted over distinct frontier members.
                byCollector.get(l.other)!.push({ edgeId: l.edgeId, other: f, t: l.t });
            }
        }
        for (const [collector, legs] of byCollector) {
            if (truncated || matches.length >= limit) return;
            if (!spend(legs.length)) return;
            const win = selectWindow(legs, st.minBranches, st.windowHours);
            if (!win) continue;
            extend(stageIdx + 1, {
                frontier: new Map([[collector, win.arrival]]),
                used: new Set([...s.used, collector]),
                edges: [...s.edges, ...win.legs.map((l) => l.edgeId)],
                layers: [...s.layers, [collector]],
            });
        }
    };

    // Stage 0 anchors the motif: a collector (fan-in) or a splitter (fan-out), one candidate per node.
    const first = stages[0];
    for (const n of g.nodes) {
        if (truncated || matches.length >= limit) break;
        const legs = (first.shape === 'fan-in' ? index[0].in : index[0].out).get(n.id) ?? [];
        if (!spend(legs.length + 1)) break;
        if (first.shape === 'fan-in' && !kindOk(n.id, first.nodeKind)) continue;
        const usable = first.shape === 'fan-out' ? legs.filter((l) => kindOk(l.other, first.nodeKind)) : legs;
        const win = selectWindow(usable, first.minBranches, first.windowHours);
        if (!win) continue;
        const others = [...new Set(win.legs.map((l) => l.other))];
        const edges = win.legs.map((l) => l.edgeId);
        if (first.shape === 'fan-in') {
            extend(1, {
                frontier: new Map([[n.id, win.arrival]]),
                used: new Set([...others, n.id]),
                edges,
                layers: [others, [n.id]],
            });
        } else {
            extend(1, {
                frontier: earliestPerNode(win.legs),
                used: new Set([n.id, ...others]),
                edges,
                layers: [[n.id], others],
            });
        }
    }
    if (matches.length >= limit) truncated = true;
    return { matches, truncated };
}

File: inspecto-geo-link/src/main/java/com/gamma/geolink/BranchingPatternEngine.java
package com.gamma.geolink;

import com.gamma.geolink.PatternQueryCompiler.Stage;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * The windowed search of a branching motif over the legs {@link PatternQueryCompiler} selected — a line-for-line
 * port of {@code matchBranchingPattern} in {@code inspecto-ui/src/app/inspecto/graph/branching-pattern-engine.ts}.
 * Keep the two in step: {@code ControlApiInvPatternTest} and {@code branching-parity.spec.ts} assert the
 * SAME golden fixture ({@code inspecto-ui/src/app/modules/admin/studio/link-analysis/branching-parity.fixture.json}).
 *
 * <p>Differences from the browser, all deliberate: there is no node cap (the SQL prune bounds the input
 * instead), and the anchors of stage 0 are visited in key order rather than projection order — only which
 * matches a {@code limit} cuts can differ, never whether a match exists.
 */
final class BranchingPatternEngine {

    /** Ceiling on leg examinations per run — the TS {@code BRANCHING_WORK_BUDGET}. */
    static final int WORK_BUDGET = 500_000;

    private BranchingPatternEngine() {}

    /** A leg of a stage: its id, endpoints (normalised keys) and time (epoch ms, null when unknown). */
    record Edge(String id, String source, String target, Long t) {}

    record Match(List<String> nodeIds, List<String> edgeIds, List<List<String>> layers) {}

    record Result(List<Match> matches, boolean truncated) {}

    /** A leg in the frame of the stage node it is grouped under: {@code other} is the counterparty. */
    private record Leg(String edgeId, String other, Long t) {}

    private record Window(List<Leg> legs, Long arrival) {}

    private record State(Map<String, Long> frontier, Set<String> used, List<String> edges, List<List<String>> layers) {}

    /** TS {@code followsInTime} (LA-14a): strictly after, within the gap; an unknown time on either side fails. */
    static boolean followsInTime(Long prev, Long t, Stage st) {
        if (!st.afterPrevious()) return true;
        if (t == null || prev == null) return false;
        if (t <= prev) return false;
        return st.maxGapHours() == null || t - prev <= st.maxGapHours() * 3_600_000;
    }

    private static Long completion(List<Leg> ordered, int min) {
        Set<String> seen = new HashSet<>();
        for (Leg l : ordered) {
            seen.add(l.other());
            if (seen.size() >= min) return l.t();
        }
        return null;
    }

    /** TS {@code selectWindow}: the EARLIEST window holding ≥ {@code min} distinct counterparties. */
    private static Window selectWindow(List<Leg> legs, int min, Double windowHours) {
        Set<String> distinct = new HashSet<>();
        for (Leg l : legs) distinct.add(l.other());
        if (distinct.size() < min) return null;
        List<Leg> timed = new ArrayList<>();
        for (Leg l : legs) if (l.t() != null) timed.add(l);
        timed.sort(Comparator.comparingLong(Leg::t));   // stable, like Array.prototype.sort
        if (windowHours == null || windowHours <= 0) return new Window(legs, completion(timed, min));
        double span = windowHours * 3_600_000;
        for (int i = 0, j = 0; i < timed.size(); i++) {
            while (j < timed.size() && timed.get(j).t() - timed.get(i).t() <= span) j++;
            List<Leg> win = timed.subList(i, j);
            Set<String> d = new HashSet<>();
            for (Leg l : win) d.add(l.other());
            if (d.size() >= min) return new Window(new ArrayList<>(win), completion(win, min));
        }
        return null;
    }

    /** TS {@code earliestPerNode}: each reached node with the EARLIEST time a counted leg reached it. */
    private static Map<String, Long> earliestPerNode(List<Leg> legs) {
        Map<String, Long> next = new LinkedHashMap<>();
        for (Leg l : legs) {
            Long prev = next.get(l.other());
            if (!next.containsKey(l.other()) || (l.t() != null && (prev == null || l.t() < prev))) next.put(l.other(), l.t());
        }
        return next;
    }

    /**
     * Find every match. {@code legsByStage.get(i)} are stage i's ELIGIBLE legs (kind + threshold already
     * applied). Every projected node is an {@code entity} — see {@link #kindOk}.
     */
    static Result match(List<Stage> stages, List<List<Edge>> legsByStage, int limit) {
        List<Map<String, List<Leg>>> outIdx = new ArrayList<>(), inIdx = new ArrayList<>();
        Set<String> nodes = new java.util.TreeSet<>();
        for (List<Edge> legs : legsByStage) {
            Map<String, List<Leg>> out = new LinkedHashMap<>(), in = new LinkedHashMap<>();
            for (Edge e : legs) {
                if (e.source().equals(e.target())) continue;
                out.computeIfAbsent(e.source(), k -> new ArrayList<>()).add(new Leg(e.id(), e.target(), e.t()));
                in.computeIfAbsent(e.target(), k -> new ArrayList<>()).add(new Leg(e.id(), e.source(), e.t()));
                nodes.add(e.source());
                nodes.add(e.target());
            }
            outIdx.add(out);
            inIdx.add(in);
        }
        Search s = new Search(stages, outIdx, limit);
        Stage first = stages.get(0);
        for (String n : nodes) {
            if (s.truncated || s.matches.size() >= limit) break;
            List<Leg> legs = (first.fanIn() ? inIdx.get(0) : outIdx.get(0)).getOrDefault(n, List.of());
            if (!s.spend(legs.size() + 1)) break;
            if (first.fanIn() && !kindOk(first.nodeKind())) continue;
            List<Leg> usable = !first.fanIn() && !kindOk(first.nodeKind()) ? List.of() : legs;
            Window win = selectWindow(usable, first.minBranches(), first.windowHours());
            if (win == null) continue;
            List<String> others = new ArrayList<>(new LinkedHashSet<>(win.legs().stream().map(Leg::other).toList()));
            List<String> edges = win.legs().stream().map(Leg::edgeId).toList();
            if (first.fanIn()) {
                Set<String> used = new LinkedHashSet<>(others);
                used.add(n);
                s.extend(1, new State(single(n, win.arrival()), used, edges, List.of(others, List.of(n))));
            } else {
                Set<String> used = new LinkedHashSet<>();
                used.add(n);
                used.addAll(others);
                s.extend(1, new State(earliestPerNode(win.legs()), used, edges, List.of(List.of(n), others)));
            }
        }
        if (s.matches.size() >= limit) s.truncated = true;
        return new Result(s.matches, s.truncated);
    }

    /** Every projected node is an {@code entity}; a stage naming another node kind matches nothing, as in the browser. */
    private static boolean kindOk(String k) {
        return k == null || PatternQueryCompiler.ENTITY_KIND.equals(k);
    }

    private static Map<String, Long> single(String node, Long at) {
        Map<String, Long> m = new LinkedHashMap<>();
        m.put(node, at);
        return m;
    }

    private static final class Search {
        final List<Stage> stages;
        final List<Map<String, List<Leg>>> out;
        final int limit;
        final List<Match> matches = new ArrayList<>();
        long work;
        boolean truncated;

        Search(List<Stage> stages, List<Map<String, List<Leg>>> out, int limit) {
            this.stages = stages;
            this.out = out;
            this.limit = limit;
        }

        boolean spend(int n) {
            work += n;
            if (work > WORK_BUDGET) truncated = true;
            return !truncated;
        }

        void extend(int stageIdx, State s) {
            if (truncated || matches.size() >= limit) return;
            if (stageIdx >= stages.size()) {
                matches.add(new Match(new ArrayList<>(s.used()), new ArrayList<>(new LinkedHashSet<>(s.edges())), s.layers()));
                return;
            }
            Stage st = stages.get(stageIdx);
            Map<String, List<Leg>> idx = out.get(stageIdx);
            if (!st.fanIn()) {
                List<Leg> legs = new ArrayList<>();
                for (Map.Entry<String, Long> f : s.frontier().entrySet()) {
                    List<Leg> cand = idx.getOrDefault(f.getKey(), List.of());
                    if (!spend(cand.size())) return;
                    for (Leg l : cand) {
                        if (s.used().contains(l.other()) || !kindOk(st.nodeKind())) continue;
                        if (followsInTime(f.getValue(), l.t(), st)) legs.add(l);
                    }
                }
                Window win = selectWindow(legs, st.minBranches(), st.windowHours());
                if (win == null) return;
                Map<String, Long> next = earliestPerNode(win.legs());
                Set<String> used = new LinkedHashSet<>(s.used());
                used.addAll(next.keySet());
                List<String> edges = new ArrayList<>(s.edges());
                win.legs().forEach(l -> edges.add(l.edgeId()));
                List<List<String>> layers = new ArrayList<>(s.layers());
                layers.add(new ArrayList<>(next.keySet()));
                extend(stageIdx + 1, new State(next, used, edges, layers));
                return;
            }
            Map<String, List<Leg>> byCollector = new LinkedHashMap<>();
            for (Map.Entry<String, Long> f : s.frontier().entrySet()) {
                List<Leg> cand = idx.getOrDefault(f.getKey(), List.of());
                if (!spend(cand.size())) return;
                for (Leg l : cand) {
                    if (s.used().contains(l.other()) || !kindOk(st.nodeKind()) || !followsInTime(f.getValue(), l.t(), st)) continue;
                    // Re-framed so `other` is the SENDER — breadth is counted over distinct frontier members.
                    byCollector.computeIfAbsent(l.other(), k -> new ArrayList<>()).add(new Leg(l.edgeId(), f.getKey(), l.t()));
                }
            }
            for (Map.Entry<String, List<Leg>> c : byCollector.entrySet()) {
                if (truncated || matches.size() >= limit) return;
                if (!spend(c.getValue().size())) return;
                Window win = selectWindow(c.getValue(), st.minBranches(), st.windowHours());
                if (win == null) continue;
                Set<String> used = new LinkedHashSet<>(s.used());
                used.add(c.getKey());
                List<String> edges = new ArrayList<>(s.edges());
                win.legs().forEach(l -> edges.add(l.edgeId()));
                List<List<String>> layers = new ArrayList<>(s.layers());
                layers.add(List.of(c.getKey()));
                extend(stageIdx + 1, new State(single(c.getKey(), win.arrival()), used, edges, layers));
            }
        }
    }
}
